### DISCUSSION TITLE
Raspberry Pico W MicroPython execution freezes a few seconds after disconnecting screen from UART

### DISCUSSION THREAD

**Discussion by cirosantilli** [micropython/micropython#9633] [Hardware & Peripherals]
I got my program running fine as explained at: https://stackoverflow.com/questions/66183596/how-can-you-make-a-micropython-program-on-a-raspberry-pi-pico-autorun/74078142#74078142

I'm installing a `main.py` that does:

```
import machine
import time

led = machine.Pin('LED', machine.Pin.OUT)
# For Rpi Pico (non-W) it was like this instead apparently.
# led = Pin(25, Pin.OUT)

i = 0
while (True):
    led.toggle()
    print(i)
    time.sleep(.5)
    i += 1
```

When I power the device on by plugging the USB to my laptop, it seems to run fine, with the LED blinking.

Then, if I connect from my laptop to the UART with:

```
screen /dev/ttyACM0 115200
```

I can see the numbers coming out on my host terminal correctly, and the LED still blinks, all as expected.

However, when I disconnect from screen with Ctrl-A K, after a few seconds, the LED stops blinking! It takes something around 15 seconds for it to stop, but it does so every time I tested.

If I reconnect the UART again with:

```
screen /dev/ttyACM0 115200
```

it starts blinking again.

Also also noticed that after I reconnect the UART and execution resumes, the count has increased much less than the actual time passed, so one possibility is that the Pico is going into some slow low power mode?

If I remove the `print()` from the program, I noticed that it does not freeze anymore after disconnecting the UART (which is of course shows no data in this case).

Micropython firmware: rp2-pico-w-20221014-unstable-v1.19.1-544-g89b320737.uf2, Ubuntu 22.04 host.

Some variants follow.

`picocom /dev/ttyACM0` instead of screen and disconnect with Ctrl-A Ctrl-Q: still freezes like with `screen`.

The following C analog of the MicroPython hacked from:

- https://github.com/raspberrypi/pico-examples/blob/a7ad17156bf60842ee55c8f86cd39e9cd7427c1d/pico_w/blink
- https://github.com/raspberrypi/pico-examples/blob/a7ad17156bf60842ee55c8f86cd39e9cd7427c1d/hello_world/usb

did not show the same problem, tested on https://github.com/raspberrypi/pico-sdk/tree/2e6142b15b8a75c1227dd3edbe839193b2bf9041

```
#include <stdio.h>
#include "pico/stdlib.h"
#include "pico/cyw43_arch.h"

int main() {
    stdio_init_all();
    if (cyw43_arch_init()) {
        printf("WiFi init failed");
        return -1;
    }
    int i = 0;
    while (true) {
        printf("%i\n", i);
        cyw43_arch_gpio_put(CYW43_WL_GPIO_LED_PIN, i % 2);
        i++;
        sleep_ms(500);
    }
    return 0;
}
```

Also asked at:

- https://stackoverflow.com/questions/74081960/raspberry-pico-w-micropython-execution-freezes-a-few-seconds-after-disconnecting

**Comment by robert-hh** [comment]
The code does not print anything. Why do you see numbers coming out?

**Reply by cirosantilli** [reply]
Sorry, copied wrong code to question, updated.

**Reply by robert-hh** [reply]
You could try the -fn option with screen, which disables flow control.

`screen -fn /dev/ttyACM0 115200`

**Reply by cirosantilli** [reply]
Thanks, tried now but still froze after disconect.

**Reply by robert-hh** [reply]
I see that here too, and with other boards as well and other terminal software. So it's not related to the Pico W, but it seems to be related to the TinyUSB stack. And, as you said, initially it works. It happens after connection and disconnecting. The time until lock-up correlates with the size of the USB output buffer.

And it does happen in a different way with a board using another USB stack (pyboard). There, the LED still blinks, but much slower.

I use picocom as terminal emulator. With the option `--no-reset`, which I usually use, this lock-up happens. Without that, it does not lock up. 

Interesting!

Nothing I can tell you at the moment to fix it, except using picocom with the command line:

`picocom /dev/ttyACM0`

**Reply by cirosantilli** [reply]
Thanks for reproducing and for the `picocom` suggestion :+1:

**Reply by cirosantilli** [reply]
`picocom /dev/ttyACM0` did freeze for me :-( I also did an analog C example, and that didn't freeze.

**Reply by robert-hh** [reply]
That's bad. So there are dependencies on the OS USB stack as well. But may I confirm that it works, if you just power up the board and do not connect?
Edit: Did you close picocom with `Ctrl-A Ctrl-X`?

**Reply by robert-hh** [reply]
> I also did an analog C example, and that didn't freeze.

The C-Driver may ignore the fact, that it cannot send data and just continue. See also the analysis here: https://github.com/micropython/micropython/issues/9634

**Reply by cirosantilli** [reply]
> That's bad. So there are dependencies on the OS USB stack as well. But may I confirm that it works, if you just power up the board and do not connect? Edit: Did you close picocom with `Ctrl-A Ctrl-X`?

Without connecting, it does not block.

I was quitting with Ctrl-A Ctrl-Q. Without Ctral-A Ctrl-X it actually does not block, so it is a valid workaround as you suggested.

Ctrl-Q does a `Skipping tty reset...` while Ctrl-A doesn't.
